src/gt4py/next/ffront/nd_array_field.py

```diff
@@ -0,0 +1,245 @@
+from numpy import typing as npt
+
+from gt4py.next import common
+
+
+try:
+    import cupy as cp
+except ImportError:
+    cp: Optional[ModuleType] = None  # type:ignore[no-redef]
+
+try:
+    from jax import numpy as jnp
+except ImportError:
+    jnp: Optional[ModuleType] = None  # type:ignore[no-redef]
+
+
+from gt4py._core import definitions
+from gt4py._core.definitions import ScalarT
+from gt4py.next.common import DimsT, DomainT
+from gt4py.next.ffront import fbuiltins
+
+
+class NonProtocolABCMeta(typing._ProtocolMeta):
+    __instancecheck__ = abc.ABCMeta.__instancecheck__
+
+
+class NonProtocolABC(metaclass=NonProtocolABCMeta):
+    pass
+
+
+def _make_unary_array_field_intrinsic_func(builtin_name: str, array_builtin_name: str) -> Callable:
+    def _builtin_unary_op(a: _BaseNdArrayField) -> common.Field:
+        xp = a.__class__.array_ns
+        op = getattr(xp, array_builtin_name)
+        new_data = op(a.ndarray)
+
+        return a.__class__.from_array(new_data, domain=a.domain)
+
+    _builtin_unary_op.__name__ = builtin_name
+    return _builtin_unary_op
+
+
+def _make_binary_array_field_intrinsic_func(builtin_name: str, array_builtin_name: str) -> Callable:
+    def _builtin_binary_op(a: _BaseNdArrayField, b: common.Field) -> common.Field:
+        xp = a.__class__.array_ns
+        op = getattr(xp, array_builtin_name)
+        if hasattr(b, "__gt_op_func__"):  # isinstance(b, common.Field):
+            if not a.domain == b.domain:
+                raise NotImplementedError(
+                    f"support for different domain not implemented: {a.domain}, {b.domain}"
+                )
+            new_data = op(a.ndarray, xp.asarray(b.ndarray))
+        else:
+            assert isinstance(b, definitions.Scalar)
+            new_data = op(a.ndarray, b)
+
+        return a.__class__.from_array(new_data, domain=a.domain)
+
+    _builtin_binary_op.__name__ = builtin_name
+    return _builtin_binary_op
+
+
+Value = Any  # TODO
+P = ParamSpec("P")
+R = TypeVar("R", Value, tuple[Value, ...])
+
+
+@dataclasses.dataclass(frozen=True)
+class _BaseNdArrayField(NonProtocolABC, common.Field[DimsT, ScalarT]):
+    _domain: DomainT
+    _ndarray: definitions.NDArrayObject
+    _value_type: ScalarT
+
+    _ops_mapping: ClassVar[dict[fbuiltins.BuiltInFunction, Callable]] = {}
+    array_ns: ClassVar[Any]  # TODO what's the better type?
+
+    @classmethod
+    def __gt_op_func__(cls, op: fbuiltins.BuiltInFunction[R, P]) -> Callable[P, R]:
+        return cls._ops_mapping.get(op, NotImplemented)
+
+    @classmethod
+    def register_gt_op_func(
+        cls, op: fbuiltins.BuiltInFunction[R, P], op_func: Optional[Callable[P, R]] = None
+    ) -> Callable[P, R]:
+        assert op not in cls._ops_mapping
+        if op_func is None:  # TODO use-case for None?
+            return functools.partial(cls.register_gt_op_func, op)
+        return cls._ops_mapping.setdefault(op, op_func)
+
+    @property
+    def domain(self) -> DomainT:
```

Comment:
If we introduce `DomainT` as a `Generic[DimsT]` we could do a bit more checks inside `Field`